### PR TITLE
Tracing for CPU

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: src/ngraph/runtime/cpu/cpu_debugger.cpp [hunk 1]
```diff
@@ -19,12 +19,26 @@
...
 using namespace std;
 using namespace ngraph;
 
+void runtime::cpu::CPU_CountTracepoint::operator()(void** outputs, const std::string& name)
+{
+    if (m_count == 0)
+    {
+        return;
+    }
+    if (++m_iteration == m_count)
```
Comment (diyessi) [inline src/ngraph/runtime/cpu/cpu_debugger.cpp]: Either do >= here, or add a clause for > and throw a runtime error if it happens.

File: src/ngraph/runtime/cpu/cpu_debugger.hpp [hunk 2]
```diff
@@ -56,13 +78,21 @@ namespace ngraph
...
+                bool
+                    add_tracepoint(std::shared_ptr<Node> op,
+                                   const std::function<void(void**, const std::string&)>& callback);
+                /// \brief Remove a tracepoint from a node
+                bool delete_tracepoint(std::shared_ptr<Node> op);
+
                 void* inspect(std::shared_ptr<Node> op, size_t output_index = 0);
 
             protected:
+                std::tuple<bool, size_t> find_pc_for_node(std::shared_ptr<Node> op);
```
Comment (diyessi) [inline src/ngraph/runtime/cpu/cpu_debugger.hpp]: Add a comment for what the two values in the tuple are.

File: src/ngraph/runtime/cpu/cpu_debugger.cpp [hunk 3]
```diff
@@ -108,3 +132,60 @@ void* runtime::cpu::CPU_Debugger::inspect(std::shared_ptr<Node> op, size_t outpu
...
+        if (replaced_functors.count(pc) != 0)
+        {
+            return false;
+        }
+
+        auto op_name = op->get_name();
+        std::vector<void**> poutputs;
+        for (size_t i = 0; i < op->get_outputs().size(); i++)
+        {
+            poutputs.push_back(&external_function->tensor_data.at(op_name + "_" + to_string(i)));
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_debugger.cpp]: Use get_tensor_data() instead. That will take care of aliases.

File: src/ngraph/runtime/cpu/cpu_debugger.cpp [hunk 4]
```diff
@@ -108,3 +132,60 @@ void* runtime::cpu::CPU_Debugger::inspect(std::shared_ptr<Node> op, size_t outpu
...
+        }
+
+        auto original_functor = external_function->functors.at(pc);
+
+        auto trace_functor =
+            [poutputs, callback, original_functor, op_name](CPURuntimeContext* ctx) {
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_debugger.cpp]: Add CPUExecutionContext*

File: src/ngraph/runtime/cpu/cpu_debugger.cpp [hunk 5]
```diff
@@ -108,3 +132,60 @@ void* runtime::cpu::CPU_Debugger::inspect(std::shared_ptr<Node> op, size_t outpu
...
+                original_functor(ctx);
+
+                std::vector<void*> outputs;
+                for (auto pout : poutputs)
+                {
+                    outputs.push_back(*pout);
+                }
+
+                callback(outputs.data(), op_name);
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_debugger.cpp]: would be even more awesome if we can reorder the data to row-major layouts (in a future PR).

File: src/ngraph/runtime/cpu/cpu_debugger.hpp [hunk 6]
```diff
@@ -56,13 +78,22 @@ namespace ngraph
...
                 void* inspect(std::shared_ptr<Node> op, size_t output_index = 0);
 
             protected:
+                // Returns a tuple with the following items <found, pc>
+                std::tuple<bool, size_t> find_pc_for_node(std::shared_ptr<Node> op);
                 CPU_Debugger(const CPU_Debugger&) = delete;
                 CPU_Debugger(CPU_Debugger&&) = delete;
                 CPU_Debugger& operator=(const CPU_Debugger&) = delete;
-
+                std::map<size_t, std::function<void(CPURuntimeContext*)>> replaced_functors;
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_debugger.hpp]: CPUExecutionContext*

File: src/ngraph/runtime/cpu/cpu_debugger.hpp [hunk 7]
```diff
@@ -33,6 +36,25 @@ namespace ngraph
     {
         namespace cpu
         {
+            class CPU_CountTracepoint
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_debugger.hpp]: Please add a comment on what callback, count and iteration are.

**Comment by raramer01** [comment]
TESTNOW

**Pull request merged**